lib/elixir/lib/protocol.ex

```diff
@@ -116,6 +116,75 @@ defmodule Protocol do
 
   @doc """
   Derives the `protocol` for `module` with the given options.
+
+  If your implementation passes options or if you are generating
+  custom code based on the struct, you will also need to provide
+  `@fallback_to_any true` or implement a macro defined as
+  `__deriving__(module, struct, options)` to get the
+  options that were passed.
+
+  ## Examples
+
+      defprotocol WithAny do
+        @fallback_to_any true
+        @doc "Ok"
+        def ok(term)
+      end
+
+      defprotocol Derivable do
+        def ok(a)
+      end
+
+      defimpl Derivable, for: Any do
+        defmacro __deriving__(module, struct, options) do
+          quote do
+            defimpl Derivable, for: unquote(module) do
+              def ok(arg) do
+                {:ok, arg, unquote(Macro.escape(struct)), unquote(options)}
+              end
+            end
+          end
+        end
+
+        def ok(arg) do
+          {:ok, arg}
+        end
+      end
+
+      defimpl WithAny, for: Any do
+        def ok(any) do
+          {:ok, any}
+        end
+      end
+
+      defmodule ImplStruct do
+        @derive [WithAny, Derivable]
+        defstruct a: 0, b: 0
+
+        defimpl Sample do
+          def ok(struct) do
+            Unknown.undefined(struct)
+          end
+        end
+      end
+
+  ## Explicit vs Implicit
+
+  Protocols can be derived implicitly and explicitly.
+  As noted above, explicit derivations can be called
+  via `@fallback_to_any true` or `__deriving__`:
+
+      # Explicitly derived via `@fallback_to_any true`
+      WithAny.ok(%ImplStruct{a: 1, b: 1})
+
+      # Explicitly derived via `__deriving__`
+      Derivable.ok(%ImplStruct{a: 1, b: 1})
+
+      # Explicitly derived by API via `__deriving__`
+      require Protocol
+      Protocol.derive(Derivable, ImplStruct, :oops)
+      Derivable.ok(%ImplStruct{a: 1, b: 1})
```

Comment:
Since they are all explicitly derived we need to rethink if we need this section since it's talking about "Explicit vs Implicit".

Reply:
i think the section is valuable but I'll remove the "Implicit vs Explicit" header

Reply:
I now feel like the documentation is wrong... according to [this test](https://github.com/elixir-lang/elixir/blob/master/lib/elixir/test/elixir/protocol_test.exs#L163) `WithAny` makes for an implicit derivation while `Derivable` makes for an explicit derivation, which is how I originally came up with the documentation. Does that mean the test descriptions are wrong?